src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs

```diff
@@ -132,11 +142,20 @@ public void Dispose()
             _logger.LogTrace("Beginning receive loop");
             try
             {
-                while (!cancellationToken.IsCancellationRequested)
+                while (await _connection.Input.WaitToReadAsync(cancellationToken))
                 {
-                    // This is a little odd... we want to remove the InvocationRequest once and only once so we pull it out in the callback,
-                    // and stash it here because we know the callback will have finished before the end of the await.
-                    var message = await _adapter.ReadMessageAsync(_stream, _binder, cancellationToken);
+                    Message incomingMessage;
+                    if (!_connection.Input.TryRead(out incomingMessage))
```

Comment:
I see. You basically want to save a call to `await _connection.Input.WaitToReadAsync`? Because even now multiple messages will work only we will have to call WaitToReadAsync for each message.